File: app/src/main/java/android/os/ServiceManager.java
package android.os;

import java.util.HashMap;
import java.util.Map;

/**
 * Stub for hidden Android API android.os.ServiceManager.
 * This class is not part of the public SDK but is available at runtime.
 */
public class ServiceManager {

    @SuppressWarnings("unchecked")
    private static Map<String, IBinder> sCache = new HashMap<>();

    public static IBinder getService(String name) {
        return sCache.get(name);
    }

    public static void addService(String name, IBinder service) {
        sCache.put(name, service);
    }

    public static String[] listServices() {
        return sCache.keySet().toArray(new String[0]);
    }
}

File: app/src/main/java/com/lmlab/voidguardian/mirror/android/os/ServiceManager.kt
package com.lmlab.voidguardian.mirror.android.os

import android.os.IBinder
import java.util.concurrent.ConcurrentHashMap

object ServiceManager {

    private val sCache = ConcurrentHashMap<String, IBinder>()

    @JvmStatic
    fun getService(name: String): IBinder? {
        return sCache[name] ?: android.os.ServiceManager.getService(name).also {
            if (it != null) sCache[name] = it
        }
    }

    fun addService(name: String, service: IBinder) {
        sCache[name] = service
    }
}
